src/shogun/neuralnets/StanNeuralNetwork.cpp

```diff
@@ -0,0 +1,280 @@
+ 	}
+
+ 	m_params = StanVector(m_total_num_parameters, 1);
+
+ 	m_params.zero();
+
+ 	for (int32_t i=0; i<m_num_layers; i++)
+ 	{
+ 		auto layer_param = get_section(m_params, i);
+
+ 		get_layer(i)->initialize_parameters(layer_param,sigma);
+
+ 		get_layer(i)->set_batch_size(m_batch_size);
+ 	}
+ }
+
+ StanNeuralNetwork::~StanNeuralNetwork()
+ {
+ 	SG_UNREF(m_layers);
+ }
+
+//TODO: Fix apply_*
+ CBinaryLabels* StanNeuralNetwork::apply_binary(CFeatures* data)
+ {
+ 	SGMatrix<float64_t> output_activations = forward_propagate(data);
+ 	CBinaryLabels* labels = new CBinaryLabels(m_batch_size);
+
+ 	for (int32_t i=0; i<m_batch_size; i++)
+ 	{
+ 		if (get_num_outputs()==1)
+ 		{
+ 			if (output_activations[i]>0.5) labels->set_label(i, 1);
+ 			else labels->set_label(i, -1);
+
+ 			labels->set_value(output_activations[i], i);
+ 		}
+ 		else if (get_num_outputs()==2)
+ 		{
+ 			float64_t v1 = output_activations[2*i];
+ 			float64_t v2 = output_activations[2*i+1];
+ 			if (v1>v2)
+ 				labels->set_label(i, 1);
+ 			else labels->set_label(i, -1);
+
+ 			labels->set_value(v2/(v1+v2), i);
+ 		}
+ 	}
+
+ 	return labels;
+ }
+
+ CRegressionLabels* StanNeuralNetwork::apply_regression(CFeatures* data)
+ {
+ 	SGMatrix<float64_t> output_activations = forward_propagate(data);
+ 	SGVector<float64_t> labels_vec(m_batch_size);
+
+ 	for (int32_t i=0; i<m_batch_size; i++)
+ 			labels_vec[i] = output_activations[i];
+
+ 	return new CRegressionLabels(labels_vec);
+ }
+
+
+ CMulticlassLabels* StanNeuralNetwork::apply_multiclass(CFeatures* data)
+ {
+ 	SGMatrix<float64_t> output_activations = forward_propagate(data);
+ 	SGVector<float64_t> labels_vec(m_batch_size);
+
+ 	for (int32_t i=0; i<m_batch_size; i++)
+ 	{
+ 		labels_vec[i] = CMath::arg_max(
+ 			output_activations.matrix+i*get_num_outputs(), 1, get_num_outputs());
+ 	}
+
+ 	CMulticlassLabels* labels = new CMulticlassLabels(labels_vec);
+
+ 	labels->allocate_confidences_for(get_num_outputs());
+ 	for (int32_t i=0; i<m_batch_size; i++)
+ 	{
+ 		labels->set_multiclass_confidences(i, SGVector<float64_t>(
+ 			output_activations.matrix, get_num_outputs(), i*get_num_outputs()));
+ 	}
+
+ 	return labels;
+ }
+
+ SGMatrix<float64_t> StanNeuralNetwork::forward_propagate(CFeatures* data, int32_t j)
+ {
+ 	SGMatrix<float64_t> inputs = features_to_matrix(data);
+ 	set_batch_size(data->get_num_vectors());
+ 	return forward_propagate(inputs, j);
+ }
+
+ SGMatrix<float64_t> StanNeuralNetwork::forward_propagate(
+ 	SGMatrix<float64_t> inputs, int32_t j)
+ {
+ 	if (j==-1)
+ 		j = m_num_layers-1;
+
+ 	for (int32_t i=0; i<=j; i++)
+ 	{
+ 		CNeuralLayer* layer = get_layer(i);
+
+ 		if (layer->is_input())
+ 			layer->compute_activations(inputs);
+ 		else
+ 			layer->compute_activations(get_section(m_params, i), m_layers);
+
+ 		layer->dropout_activations();
+ 	}
+
+ 	return get_layer(j)->get_activations();
+ }
+
+
+ SGVector<float64_t>* StanNeuralNetwork::get_layer_parameters(int32_t i)
```

Comment:
This looks strange. The signature in the header is `StanVector& get_layer_parameters(int32_t i);`.